### PR TITLE
Rdg 40 create enemy mechanics

### PR DESCRIPTION
**Pull request by MatthijsLuijten**
## Describe your changes
Added an `Enemy` class that enherits from the `Agent` class. Added the functionality in the main loop and environment.
Also, in the pre-commit I added a hook that automatically runs all tests.

## List of tests added
`test_initialization`
`test_move`
`test_render`
there is not `test_reset` added because this is not done for the enemies.

## Checklist before requesting a review
- [x] PR title contains ticket ID
- [x] If it is a core feature, I have added thorough tests.
- [x] I have included documentation

### CODE REVIEW

File: src/environment.py [hunk 1]
```diff
@@ -99,11 +104,37 @@ def move_player(self, action: int) -> Tuple[int, bool]:
...
+        # Check if the player collides with an enemy
+        for enemy in self.enemies:
+            if self.check_collision(self.player, enemy):
+                reward = self.rewards["death"]
+                done = True
+                return reward, done
 
         return reward, done
 
+    def check_collision(self, player: Player, enemy: Enemy) -> bool:
```
Comment (DaphneSmits) [inline src/environment.py]: We could also make this function a bit more generic by accepting two `Agent`s as input and computing the collision between these agents (without restricting it to be exactly a `Player` and an `Enemy`). This way we could also use this function in the future to check collisions between two enemies or two players.
Just a thought, but maybe we don't really want that and only want to use this for this specific case and therefore also keep it specific.

File: src/environment.py [hunk 2]
```diff
@@ -141,10 +172,77 @@ def step(self, action: int) -> Tuple[int, np.ndarray, bool]:
...
+            # This ensures that self.enemies is not getting infinitely large
+            if self.is_moving_outside_the_screen(enemy):
+                self.enemies.remove(enemy)
 
         if self.render_on:
             self.render()
 
         return reward, next_state, done
+
+    def is_moving_outside_the_screen(self, enemy: Enemy) -> bool:
```
Comment (DaphneSmits) [inline src/environment.py:189]: I feel like this function is very similar to the `is_valid_location()` function, as both check if an agent is going outside the screen. Maybe you could update that function to be more generic and reuse it for the enemies as well? I do like this name, it is more descriptive than `is_valid_location()`.

**Review by DaphneSmits — COMMENTED** [review]
Nice work!! I added some minor comments about making the code as generic as possible. I am curious what you think. Also please say if you think my suggestions will just make it unnecessary complex!

**Comment by MatthijsLuijten** [comment]
@DaphneSmits I addressed your comments

**Pull request merged**